core/storage/couchStorageHelpers.go

```diff
@@ -0,0 +1,582 @@
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) upsertObject(id string, object interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	if _, err := db.Put(context.TODO(), id, object); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) getInstanceID() int64 {
+	return time.Now().UnixNano() / (int64(time.Millisecond) / int64(time.Nanosecond))
+}
+
+func (store *CouchStorage) findAll(query interface{}, result interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return notFound
+		}
+		return err
+	}
+
+	resultv := reflect.ValueOf(result)
+	if resultv.Kind() != reflect.Ptr {
+		return &Error{"result argument must be a slice address"}
+	}
+
+	slicev := resultv.Elem()
+
+	if slicev.Kind() == reflect.Interface {
+		slicev = slicev.Elem()
+	}
+	if slicev.Kind() != reflect.Slice {
+		return &Error{"result argument must be a slice address"}
+	}
+
+	slicev = slicev.Slice(0, slicev.Cap())
+	elemt := slicev.Type().Elem()
+	i := 0
+	for rows.Next() {
+		elemp := reflect.New(elemt)
+		err := rows.ScanDoc(elemp.Interface())
+		if err != nil {
+			return err
+		}
+
+		slicev = reflect.Append(slicev, elemp.Elem())
+		slicev = slicev.Slice(0, slicev.Cap())
+		i++
+	}
+	resultv.Elem().Set(slicev.Slice(0, i))
+
+	if err := rows.Err(); err != nil {
+		return err
+	}
+	return rows.Close()
+}
+
+func (store *CouchStorage) countObjects(query interface{}) (uint32, common.SyncServiceError) {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return 0, nil
+		}
+		return 0, err
+	}
+
+	var count uint32
+	for rows.Next() {
+		count++
+	}
+
+	if err := rows.Err(); err != nil {
+		return 0, err
+	}
+
+	return count, nil
+}
+
+func (store *CouchStorage) deleteObject(id string) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+	row := db.Get(context.TODO(), id)
+	if kivik.StatusCode(row.Err) == http.StatusNotFound {
+		return notFound
+	}
+
+	if _, err := db.Delete(context.TODO(), id, row.Rev); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (store *CouchStorage) deleteAllCouchObjects(query interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	rows, err := db.Find(context.TODO(), query)
+	if err != nil {
+		if kivik.StatusCode(err) == http.StatusNotFound {
+			return notFound
+		}
+		return err
+	}
+
+	for rows.Next() {
+
+		var tempObject couchObject
+		err = rows.ScanDoc(&tempObject)
+		if err != nil {
+			return err
+		}
+
+		if _, err := db.Delete(context.TODO(), tempObject.ID, tempObject.Rev); err != nil {
```

Comment:
Similar comment as on the question of counting rows. Is there a way to pass a query string to a delete function and have it delete everything that matches without having to return the rows to you so that you can delete them one at a time? This comment applies to all the delete all functions in the storage interface.

Reply:
I did research about this as well, there is no in-built way to delete all objects that satisfy a particular query. However, there is a way to 'update' objects in bulk, which means I could delete all objects in one call. But I would still have to call Find to get all objects that satisfy the query. In case of bulk deleting, I would need to pass a slice of interfaces, with each interface in the slice containing the 'id', 'rev' and '_deleted = true' for the object to be deleted. This bulk update function would then return a result set with the status of each object, so I would have to iterate over that to check for errors corresponding to each. Do you think this would be better in terms of performance?

Reply:
ok. I'm not sure it's worth doing that. Actually, it makes sinse with the way couch works that you have to "Get" the object so that you can get the rev, because you need the rev to delete it. Make of note of this behavior in the design document. This is an important finding about how couch works so it should be called out. Also, consider a good spot in the code to place a comment, few sentences long about this behavior.